fe/fe-core/src/main/java/org/apache/doris/nereids/cost/StatsCalculator.java

```diff
@@ -0,0 +1,101 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.cost;
+
+import org.apache.doris.nereids.PlanContext;
+import org.apache.doris.nereids.memo.GroupExpression;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalAggregate;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalDistribution;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalFilter;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalHashJoin;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalHeapSort;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalOlapScan;
+import org.apache.doris.nereids.trees.plans.physical.PhysicalProject;
+import org.apache.doris.nereids.trees.plans.visitor.DefaultPlanVisitor;
+import org.apache.doris.statistics.AggStatsDerive;
+import org.apache.doris.statistics.ExchangeStatsDerive;
+import org.apache.doris.statistics.FilterStatsDerive;
+import org.apache.doris.statistics.HashJoinStatsDerive;
+import org.apache.doris.statistics.OlapScanStatsDerive;
+import org.apache.doris.statistics.SelectStatsDerive;
+import org.apache.doris.statistics.StatsDeriveResult;
+
+/**
+ * This class is used to get the statistics info for each group.
+ */
+public class StatsCalculator extends DefaultPlanVisitor<StatsDeriveResult, PlanContext> {
+
+    /**
+     * Gather statistic information of GroupExpression's parent group.
+     */
+    public void estimate(GroupExpression groupExpression) {
+        Plan plan = groupExpression.getPlan();
```

Comment:
groupExpression.getPlan() will return a plan with GroupPlan as children, so DefaultPlanVisitor will stop recursive visit plan when visit GroupPlan.

so you must rewrite visitGroupPlan to recursive visit children plans

Reply:
I think there is no need to recursively visit child plan, DeriveStatsJob will handle this recursively with a stack.